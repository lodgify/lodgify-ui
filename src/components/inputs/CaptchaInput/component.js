import React from 'react';
import PropTypes from 'prop-types';
import { Form, Image } from 'semantic-ui-react';

import { TextInput } from 'inputs/TextInput';

/**
 * A captcha input helps to ensure a user is a real human.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  value,
  error,
  image,
  isValid,
  label,
  name,
  onBlur,
  onChange,
}) => (
  <Form.Group widths="equal">
    <Form.Field>
      <Image src={image} />
    </Form.Field>
    <Form.Field>
      <TextInput
        autoComplete="off"
        error={error}
        isValid={isValid}
        label={label}
        name={name}
        onBlur={onBlur}
        onChange={onChange}
        value={value}
      />
    </Form.Field>
  </Form.Group>
);

Component.displayName = 'CaptchaInput';

Component.defaultProps = {
  error: false,
  isValid: false,
  label: '',
  name: '',
  onBlur: Function.prototype,
  onChange: Function.prototype,
  value: undefined,
};

Component.propTypes = {
  /** Is the captcha input in an error state. */
  error: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  /** The source url for the image to display. */
  image: PropTypes.string.isRequired,
  /** Is the captcha input in a valid state. */
  isValid: PropTypes.bool,
  /** The visible label for the text input. */
  label: PropTypes.string,
  /** The name for the captcha input. */
  name: PropTypes.string,
  /**
   * Used internally by `Form` so ignored in the styleguide.
   * @ignore
   */
  onBlur: PropTypes.func,
  /**
   * A function called when the captcha input value changes
   * @param {String} name
   * @param {String} value
   */
  onChange: PropTypes.func,
  /** The current value of the input where it is consumed as a controlled component. */
  value: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.string,
    PropTypes.number,
  ]),
};
