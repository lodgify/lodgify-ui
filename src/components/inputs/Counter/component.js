import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { FlexContainer } from 'layout/FlexContainer';
import { Button } from 'elements/Button';
import { ICON_NAMES } from 'elements/Icon';
import { Paragraph } from 'typography/Paragraph';
import { getControlledInputValue } from 'utils/get-controlled-input-value';

/**
 * A counter allows a user to select a number value in increments of one.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export class Component extends PureComponent {
  state = {
    value: this.props.initialValue || 0,
  };

  componentDidUpdate = (previousProps, previousState) => {
    const { value } = this.state;

    if (value === previousState.value) return;

    const { name, onChange } = this.props;

    onChange(name, value);
  };

  handleDecrement = () => {
    const value = getControlledInputValue(
      this.props.value,
      undefined,
      this.state.value
    );

    value !== 0 && this.setState({ value: value - 1 });
  };

  handleIncrement = () => {
    const value = getControlledInputValue(
      this.props.value,
      undefined,
      this.state.value
    );

    value !== this.props.max && this.setState({ value: value + 1 });
  };

  render = () => {
    const value = getControlledInputValue(
      this.props.value,
      undefined,
      this.state.value
    );

    return (
      <div className="ui counter">
        <FlexContainer alignItems="center" justifyContent="space-between">
          <Button
            icon={ICON_NAMES.MINUS}
            isOutlined
            isRounded
            onClick={this.handleDecrement}
          />
          <Paragraph>{this.props.renderValue(value)}</Paragraph>
          <Button
            icon={ICON_NAMES.PLUS}
            isOutlined
            isRounded
            onClick={this.handleIncrement}
          />
        </FlexContainer>
      </div>
    );
  };
}

Component.displayName = 'Counter';

Component.defaultProps = {
  initialValue: undefined,
  max: undefined,
  name: '',
  onChange: Function.prototype,
  renderValue: value => value,
  value: undefined,
};

Component.propTypes = {
  /** The default value of the range. Minimum 0. */
  initialValue: PropTypes.number,
  /** The maximum value the input accepts. */
  max: PropTypes.number,
  /** The name for the range. */
  name: PropTypes.string,
  /** A function called when the range value changes. */
  onChange: PropTypes.func,
  /** A function returning a React node for modifying the way values are rendered. */
  renderValue: PropTypes.func,
  /** The current value of the range where it is consumed as a controlled component. Minimum 0. */
  value: PropTypes.number,
};
