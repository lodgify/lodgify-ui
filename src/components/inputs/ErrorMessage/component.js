import 'semantic-ui-styles/label.less';
import React from 'react';
import PropTypes from 'prop-types';
import { Label } from 'semantic-ui-react';

/**
 * Error message for use with input elements
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({ errorMessage }) => (
  <Label color="red" pointing="below">
    {errorMessage}
  </Label>
);

Component.displayName = 'ErrorMessage';

Component.propTypes = {
  /** The error message to display */
  errorMessage: PropTypes.string.isRequired,
};
