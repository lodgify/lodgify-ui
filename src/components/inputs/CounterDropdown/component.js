import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Popup } from 'semantic-ui-react';
import classnames from 'classnames';

import { some } from 'utils/some';
import { GUESTS } from 'utils/default-strings';
import { Counter } from 'inputs/Counter';
import { Icon, ICON_NAMES } from 'elements/Icon';
import { FlexContainer } from 'layout/FlexContainer';
import { testidFactory } from 'utils/testid';

const TEST_ID_PREFIX = 'counter-dropdown';

const testid = testidFactory(TEST_ID_PREFIX);

/**
 * A counter dropdown allows a user to select a value from a counter input.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  counterName,
  counterValue,
  dropdownLabel,
  maximumCounterValue,
  onChange,
  willOpenAbove,
}) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <Popup
      className="counter-dropdown"
      content={
        <FlexContainer flexDirection="row" justifyContent="space-between">
          <Counter
            max={maximumCounterValue}
            name={counterName}
            onChange={onChange}
            value={counterValue}
          />
        </FlexContainer>
      }
      data-testid={testid()}
      on="click"
      onClose={() => setIsOpen(false)}
      onOpen={() => setIsOpen(true)}
      open={isOpen}
      position={willOpenAbove ? 'top left' : 'bottom left'}
      text={dropdownLabel}
      trigger={
        <div
          className={classnames('dropdown-container', {
            dirty: some(counterValue),
          })}
          role="button"
        >
          <div
            className={classnames(
              'ui selection dropdown',
              'counter-dropdown-trigger',
              {
                active: isOpen,
              }
            )}
          >
            <span>{dropdownLabel}</span>
            <Icon name={ICON_NAMES.CARET_DOWN}></Icon>
          </div>
        </div>
      }
    />
  );
};

Component.displayName = 'CounterDropdown';

Component.defaultProps = {
  counterName: undefined,
  counterValue: undefined,
  dropdownLabel: GUESTS,
  maximumCounterValue: undefined,
  onChange: undefined,
  willOpenAbove: false,
};

Component.propTypes = {
  /** The name for the counter. */
  counterName: PropTypes.string,
  /** The value of the counter. Minimum 0. */
  counterValue: PropTypes.number,
  /** The text that populates the dropdown's label. */
  dropdownLabel: PropTypes.string,
  /** The maximum value the counter accepts. */
  maximumCounterValue: PropTypes.number,
  /** A function called when the counter value changes. */
  onChange: PropTypes.func,
  /** Will the dropdown open above or below the trigger */
  willOpenAbove: PropTypes.bool,
};
