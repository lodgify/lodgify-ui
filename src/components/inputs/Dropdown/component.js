import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Dropdown } from 'semantic-ui-react';
import getClassNames from 'classnames';

import { getIsInputValueReset } from 'utils/get-is-input-value-reset';
import { getHasErrorMessage } from 'utils/get-has-error-message';
import { Icon, ICON_NAMES } from 'elements/Icon';
import { NO_RESULTS } from 'utils/default-strings';
import { getControlledInputValue } from 'utils/get-controlled-input-value';

import { ErrorMessage } from '../ErrorMessage';

import { getIsOpenAfterChange } from './utils/getIsOpenAfterChange';
import { adaptOptions } from './utils/adaptOptions';
import { getDefaultValue } from './utils/getDefaultValue';
import { getHasImages } from './utils/getHasImages';
import { isValueValid } from './utils/isValueValid';

/**
 * A dropdown allows a user to select a value from a series of options.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export class Component extends PureComponent {
  state = {
    isOpen: false,
    value: undefined,
  };

  componentDidUpdate(previousProps, previousState) {
    if (getIsInputValueReset(previousProps.value, this.props.value)) {
      this.setState({ value: undefined });
      return;
    }

    const { value } = this.state;
    const { name, onChange } = this.props;

    previousState.value !== value && onChange(name, value);
  }

  handleChange = ({ key }, { value }) => {
    this.setState({
      value,
      isOpen: getIsOpenAfterChange(key),
    });
  };

  handleOpen = isOpen => this.setState({ isOpen });

  handleBlur = (event, data) => {
    this.props.onBlur(event, data);
    this.handleOpen(false);
  };

  render() {
    const { isOpen } = this.state;
    const {
      error,
      icon,
      isDisabled,
      isSearchable,
      isValid,
      noResultsText,
      label,
      options,
      willOpenAbove,
    } = this.props;
    const value = getControlledInputValue(
      this.props.value,
      undefined,
      this.state.value
    );
    const hasImages = getHasImages(options);
    const adaptedOptions = adaptOptions(options, hasImages);
    const hasErrorMessage = getHasErrorMessage(error);

    return (
      <div
        className={getClassNames('dropdown-container', {
          'has-images': hasImages,
          dirty: isValueValid(value),
          error: error,
          focus: isOpen,
          valid: isValid,
        })}
      >
        {hasErrorMessage && <ErrorMessage errorMessage={error} />}
        {isValid && <Icon color="green" name={ICON_NAMES.CHECKMARK} />}
        {!hasImages && icon && <Icon name={icon} />}
        <Dropdown
          defaultValue={getDefaultValue(
            adaptedOptions,
            hasImages,
            !!label,
            value
          )}
          disabled={isDisabled || !adaptedOptions.length}
          icon={<Icon name={ICON_NAMES.CARET_DOWN} />}
          noResultsMessage={noResultsText}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
          onClick={() => this.handleOpen(!isOpen)}
          open={isOpen}
          options={adaptedOptions}
          search={isSearchable}
          selectOnBlur={false}
          selection
          upward={willOpenAbove}
          value={value}
        />
        {!hasImages && label && (
          <label onClick={() => this.handleOpen(true)}>{label}</label>
        )}
      </div>
    );
  }
}

Component.displayName = 'Dropdown';

Component.defaultProps = {
  error: false,
  icon: null,
  isDisabled: false,
  isSearchable: false,
  isValid: false,
  label: '',
  name: '',
  noResultsText: NO_RESULTS,
  onBlur: Function.prototype,
  onChange: Function.prototype,
  options: [],
  value: undefined,
  willOpenAbove: false,
};

Component.propTypes = {
  /** Is the dropdown in an error state. */
  error: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  /** Icon for the dropdown. Not displayed if options have images. */
  icon: PropTypes.string,
  /** A disabled dropdown does not allow user interaction. */
  isDisabled: PropTypes.bool,
  /** Is the dropdown searchable with keyboard input */
  isSearchable: PropTypes.bool,
  /** Is the dropdown in a valid state. */
  isValid: PropTypes.bool,
  /** The label for the dropdown. Not displayed if options have images. */
  label: PropTypes.string,
  /** The name for the dropdown. */
  name: PropTypes.string,
  /** The text to display when no results are returned from a searchable dropdown. */
  noResultsText: PropTypes.string,
  /**
   * Used internally by `Form` so ignored in the styleguide.
   * @ignore
   */
  onBlur: PropTypes.func,
  /** A function called when the dropdown value changes. */
  onChange: PropTypes.func,
  /** The options which the user can select. Dropdown is disabled if options is an empty array. */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      /** A list of one or more strings separated by commas indicating a set of source sizes for the image. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
      imageSizes: PropTypes.string,
      /** A list of one or more strings separated by commas indicating a set of possible image sources for the user agent to use for the image. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
      imageSrcSet: PropTypes.string,
      /** The source url of the image. */
      imageUrl: PropTypes.string,
      /** The indent level of an option. One of: 0, 1, 2, 3, 4, 5 */
      indent: PropTypes.oneOf([0, 1, 2, 3, 4, 5]),
      /** The visible label for the option. */
      label: PropTypes.string,
      /** The visible text for the option. */
      text: PropTypes.string.isRequired,
      /** The underlying value for the option. */
      value: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.number,
        PropTypes.string,
      ]),
    })
  ),
  /** The current value of the dropdown where it is consumed as a controlled component. */
  value: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.number,
    PropTypes.string,
  ]),
  /** Should the dropdown display upward */
  willOpenAbove: PropTypes.bool,
};
