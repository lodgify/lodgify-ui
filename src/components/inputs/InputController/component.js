import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { isEqual, some } from 'lodash';
import { Input } from 'semantic-ui-react';
import getClassNames from 'classnames';

import { getIsInputValueReset } from 'utils/get-is-input-value-reset';
import { getControlledInputValue } from 'utils/get-controlled-input-value';
import { getHasErrorMessage } from 'utils/get-has-error-message';
import { Icon, ICON_NAMES } from 'elements/Icon';

import { ErrorMessage } from '../ErrorMessage';

import { getValueFromEvent } from './utils/getValueFromEvent';

/**
 * Shared controller for input elements.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export class Component extends PureComponent {
  state = {
    value: this.props.initialValue,
  };

  componentDidUpdate(previousProps, previousState) {
    if (getIsInputValueReset(previousProps.value, this.props.value)) {
      this.setState({ value: this.props.initialValue });
      return;
    }

    const { value } = this.state;
    const { name, onChange } = this.props;

    !isEqual(previousState.value, value) && onChange(name, value);
  }

  handleChange = (...args) => {
    const value = this.props.adaptOnChangeEvent(...args);

    this.setState({ value });
  };

  /**
   * Force focus the input when the dynamic label is clicked.
   */
  handleClick = () => this.htmlInput.focus();

  render() {
    const {
      children,
      error,
      icon,
      initialValue,
      inputOnChangeFunctionName,
      isFocused,
      isValid,
      label,
      mapValueToProps,
    } = this.props;
    const value = getControlledInputValue(
      this.props.value,
      initialValue,
      this.state.value
    );
    const isDirty = some(value);
    const hasErrorMessage = getHasErrorMessage(error);

    return (
      <Input
        className={getClassNames({
          dirty: isDirty,
          valid: isValid,
          error: error,
          focus: isFocused,
        })}
        iconPosition={icon && 'left'}
      >
        {hasErrorMessage && <ErrorMessage errorMessage={error} />}
        {isValid && <Icon color="green" name={ICON_NAMES.CHECKMARK} />}
        {React.cloneElement(children, {
          [inputOnChangeFunctionName]: this.handleChange,
          ref: input => (this.htmlInput = input),
          ...mapValueToProps(value),
        })}
        {label && <label onClick={this.handleClick}>{label}</label>}
        {icon}
      </Input>
    );
  }
}

Component.displayName = 'InputController';

Component.defaultProps = {
  adaptOnChangeEvent: getValueFromEvent,
  icon: null,
  initialValue: '',
  inputOnChangeFunctionName: 'onChange',
  isFocused: false,
  label: null,
  mapValueToProps: value => ({ value }),
  value: undefined,
};

Component.propTypes = {
  /** A function called to adapt the child input `onChange` event to a value.  */
  adaptOnChangeEvent: PropTypes.func,
  /** The input controlled by the input controller */
  children: PropTypes.element.isRequired,
  /** Is input in an error state. */
  error: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]).isRequired,
  /** An icon to display in the input */
  icon: PropTypes.element,
  /** The value of the input on first render when it is consumed as a controlled component. */
  // eslint-disable-next-line react/forbid-prop-types
  initialValue: PropTypes.any,
  /** The name of the change function on the input, if not `onChange`.  */
  inputOnChangeFunctionName: PropTypes.string,
  /** Is input in a focused state. */
  isFocused: PropTypes.bool,
  /** Is input in a valid state. */
  isValid: PropTypes.bool.isRequired,
  /** The visible label for the input. */
  label: PropTypes.string,
  /**
   * A function which maps the value persisted in the input controller state to the props passed to the input.
   * @param  {any} value
   * @return {Object}
   */
  mapValueToProps: PropTypes.func,
  /** The name for the input. */
  name: PropTypes.string.isRequired,
  /**
   * A function called when a change in the input value is handled by the input controller.
   * @param {String} name
   * @param {String} value
   */
  onChange: PropTypes.func.isRequired,
  /** The current value of the input where it is consumed as a controlled component. */
  // eslint-disable-next-line react/forbid-prop-types
  value: PropTypes.any,
};
