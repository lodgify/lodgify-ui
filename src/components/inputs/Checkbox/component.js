import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from 'semantic-ui-react';

import { InputController } from '../InputController';

import { mapValueToProps } from './utils/mapValueToProps';
import { adaptOnChangeEvent } from './utils/adaptOnChangeEvent';

/**
 * A checkbox allows a user to select a value from a small set of options, often binary.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  error,
  isDisabled,
  isChecked,
  isRadioButton,
  isToggle,
  isValid,
  label,
  name,
  onChange,
}) => (
  <InputController
    adaptOnChangeEvent={adaptOnChangeEvent}
    error={error}
    isValid={isValid}
    label={label}
    mapValueToProps={mapValueToProps}
    name={name}
    onChange={onChange}
    value={isChecked}
  >
    <Checkbox disabled={isDisabled} radio={isRadioButton} toggle={isToggle} />
  </InputController>
);

Component.displayName = 'Checkbox';

Component.defaultProps = {
  error: false,
  isChecked: undefined,
  isDisabled: false,
  isToggle: false,
  isRadioButton: false,
  isValid: false,
  label: '',
  name: '',
  onChange: Function.prototype,
};

Component.propTypes = {
  /** Is the checkbox input in an error state. */
  error: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  /** Is the checkbox checked. */
  isChecked: PropTypes.bool,
  /** Is the checkbox disabled. */
  isDisabled: PropTypes.bool,
  /** Is the checkbox formatted as a radio button. Prefer using the <RadioButton> component instead */
  isRadioButton: PropTypes.bool,
  /** Format to show an on or off choice. */
  isToggle: PropTypes.bool,
  /** Is the checkbox in a valid state. */
  isValid: PropTypes.bool,
  /** The label for the checkbox */
  label: PropTypes.node,
  /** The HTML input name. */
  name: PropTypes.string,
  /**
   * Event called when the checkbox value changes.
   * @param {String} name
   * @param{boolean} value
   */
  onChange: PropTypes.func,
};
