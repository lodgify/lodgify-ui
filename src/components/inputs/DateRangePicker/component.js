import moment from 'moment';
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { uniqueId, debounce } from 'lodash';
import { DateRangePicker } from 'react-dates';

import 'react-dates/initialize';

import { getUpOrDownFromBoolean } from 'utils/get-up-or-down-from-boolean';
import { withResponsive } from 'utils/with-responsive';
import { Icon, ICON_NAMES } from 'elements/Icon';
import { InputController } from 'inputs/InputController';
import { isBlurEvent } from 'utils/is-blur-event';
import { returnFirstArgument } from 'utils/return-first-argument';
import { getWindowHeight } from 'utils/get-window-height';
import { isDisplayedAsModal } from 'utils/is-displayed-as-modal';

import { mapValueToProps } from './utils/mapValueToProps';
import { getNumberOfMonths } from './utils/getNumberOfMonths';
import { MAXIMUM_SCREEN_WIDTH_FOR_TWO_MONTH_CALENDAR } from './constants';

/**
 * A date range picker lets a user pick a date range.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
class Component extends PureComponent {
  state = {
    focusedInput: null,
    windowHeight: null,
  };

  componentDidMount = () => {
    moment.locale(this.props.localeCode);
    this.handleHeightChange();
    global.addEventListener('resize', debounce(this.handleHeightChange, 150));
  };

  componentDidUpdate = (prevProps, prevState) => {
    const { focusedInput: prevFocusedInput } = prevState;
    const { focusedInput } = this.state;
    const { onBlur } = this.props;

    isBlurEvent(prevFocusedInput, focusedInput) && onBlur();
  };

  handleFocusChange = focusedInput => {
    this.setState({ focusedInput });
  };

  handleHeightChange = () => {
    const windowHeight = getWindowHeight();

    if (windowHeight !== this.state.windowHeight) {
      this.setState({
        windowHeight,
      });
    }
  };

  render() {
    const {
      displayFormat,
      endDatePlaceholderText,
      error,
      getIsDayBlocked,
      isValid,
      name,
      onChange,
      startDatePlaceholderText,
      value,
      willOpenAbove,
      windowInnerWidth,
    } = this.props;
    const { focusedInput } = this.state;

    return (
      <InputController
        adaptOnChangeEvent={returnFirstArgument}
        error={error}
        initialValue={null}
        inputOnChangeFunctionName="onDatesChange"
        isFocused={!!focusedInput}
        isValid={isValid}
        mapValueToProps={mapValueToProps}
        name={name}
        onChange={onChange}
        value={value}
      >
        <DateRangePicker
          /* eslint-disable react/jsx-sort-props */
          // Consumer defined props.
          displayFormat={displayFormat}
          endDatePlaceholderText={endDatePlaceholderText}
          isDayBlocked={getIsDayBlocked}
          openDirection={getUpOrDownFromBoolean(willOpenAbove)}
          startDatePlaceholderText={startDatePlaceholderText}
          // Controlled props
          focusedInput={focusedInput}
          // NOTE onDatesChange is required by DateRangePicker but is set in `InputController`
          onDatesChange={Function.prototype}
          onFocusChange={this.handleFocusChange}
          // Static required props.
          endDateId={uniqueId('end_date_id_')}
          startDateId={uniqueId('start_date_id_')}
          // Static custom appearance props.
          customArrowIcon={<Icon name={ICON_NAMES.ARROW_RIGHT} />}
          customInputIcon={<Icon name={ICON_NAMES.CALENDAR} />}
          daySize={52}
          hideKeyboardShortcutsPanel
          navNext={<Icon name={ICON_NAMES.ARROW_RIGHT} />}
          navPrev={<Icon name={ICON_NAMES.ARROW_LEFT} />}
          numberOfMonths={getNumberOfMonths(windowInnerWidth)}
          withPortal={isDisplayedAsModal(this.state.windowHeight)}
          /* eslint-enable react/jsx-sort-props */
        />
      </InputController>
    );
  }
}

Component.displayName = 'DateRangePicker';

Component.defaultProps = {
  displayFormat: 'DD/MM/YYYY',
  endDatePlaceholderText: '',
  error: false,
  getIsDayBlocked: Function.prototype,
  isValid: false,
  localeCode: 'en',
  name: '',
  onBlur: Function.prototype,
  onChange: Function.prototype,
  startDatePlaceholderText: '',
  value: undefined,
  willOpenAbove: false,
  windowInnerWidth: MAXIMUM_SCREEN_WIDTH_FOR_TWO_MONTH_CALENDAR,
};

Component.propTypes = {
  /** The format for inputting and displaying dates as text. */
  displayFormat: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  /** The visible placeholder text for the end date input. */
  endDatePlaceholderText: PropTypes.string,
  /** Is the date range picker in an error state. A string is displayed as an error message. */
  error: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  /**
   * A function called for each day to be displayed. Returning true blocks that day in the date range picker.
   * @param   {Moment}  day - The day to test.
   * @returns {boolean}     - Is the day blocked.
   */
  getIsDayBlocked: PropTypes.func,
  /** Is the date range picker in a valid state. */
  isValid: PropTypes.bool,
  /** The ISO 639-1 locale code which changes the format and language of days of the week and the months of the year. */
  localeCode: PropTypes.string,
  /** The name for the date range picker. */
  name: PropTypes.string,
  /**
   * Used internally by `Form` so ignored in the styleguide.
   * @ignore
   */
  onBlur: PropTypes.func,
  /**
   * A function called when the date range picker value changes. Dates are wrapped as [Moment objects](https://momentjs.com/docs/#/parsing/).
   * @param {String} name   - `this.props.name`
   * @param {Object} value
   * @param {Moment} value.endDate
   * @param {Moment} value.startDate
   */
  onChange: PropTypes.func,
  /** The visible placeholder text for the start date input. */
  startDatePlaceholderText: PropTypes.string,
  /** The current value of the input where it is consumed as a controlled component. */
  value: PropTypes.shape({
    endDate: PropTypes.object,
    startDate: PropTypes.object,
  }),
  /** Will the calendar open above the input. */
  willOpenAbove: PropTypes.bool,
  /**
   * Is the user on a mobile device.
   * Provided by `withResponsive` so ignored in the styleguide.
   * @ignore
   */
  windowInnerWidth: PropTypes.number,
};

export const ComponentWithResponsive = withResponsive(Component);
