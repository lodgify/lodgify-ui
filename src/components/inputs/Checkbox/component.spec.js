import React from 'react';
import { shallow } from 'enzyme';
import { Checkbox as SemanticCheckbox } from 'semantic-ui-react';
import {
  expectComponentToBe,
  expectComponentToHaveDisplayName,
  expectComponentToHaveProps,
} from '@lodgify/enzyme-jest-expect-helpers';

import { Component as Checkbox } from './component';

const getCheckbox = () => shallow(<Checkbox />);

describe('<Checkbox />', () => {
  it('should render a single Semantic UI Checkbox component', () => {
    const wrapper = getCheckbox();

    expectComponentToBe(wrapper, SemanticCheckbox);
  });

  describe('`SemanticCheckbox`', () => {
    it('should have the right props', () => {
      const wrapper = getCheckbox().find(SemanticCheckbox);

      expectComponentToHaveProps(wrapper, {
        checked: undefined,
        defaultChecked: undefined,
        disabled: false,
        label: '',
        name: '',
        onChange: expect.any(Function),
        radio: false,
        toggle: false,
      });
    });
  });

  describe('Interaction: onChange', () => {
    it('should call `props.onChange` with the right arguments', () => {
      const name = 'someName';
      const handleChange = jest.fn();
      const checkbox = shallow(
        <Checkbox name={name} onChange={handleChange} />
      );

      checkbox.simulate('change', undefined, { checked: true });
      expect(handleChange).toHaveBeenCalledWith(name, true);
    });
  });

  it('should have displayName "Checkbox"', () => {
    expectComponentToHaveDisplayName(Checkbox, 'Checkbox');
  });
});
