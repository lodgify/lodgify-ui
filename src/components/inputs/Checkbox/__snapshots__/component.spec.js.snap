// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> if \`props.isFluid\` is true should return the right structure 1`] = `
<Checkbox
  error={false}
  isCompact={false}
  isDisabled={false}
  isFluid={true}
  isLabelLeft={false}
  isRadioButton={false}
  isToggle={false}
  label=""
  name=""
  onChange={[Function]}
  onClick={[Function]}
>
  <InputController
    adaptOnChangeEvent={[Function]}
    error={false}
    icon={null}
    initialValue=""
    inputOnChangeFunctionName="onChange"
    isCompact={false}
    isFluid={true}
    isFocused={false}
    isValid={false}
    label={null}
    mapValueToProps={[Function]}
    name=""
    onChange={[Function]}
  >
    <Input
      className=""
      fluid={true}
      iconPosition={null}
      type="text"
    >
      <div
        className="ui fluid input"
      >
        <Checkbox
          checked={false}
          className="is-fluid"
          disabled={false}
          key=".0"
          label=""
          onChange={[Function]}
          onClick={[Function]}
          radio={false}
          toggle={false}
          type="checkbox"
        >
          <div
            className="ui checkbox is-fluid"
            onChange={[Function]}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
          >
            <Ref
              innerRef={
                Object {
                  "current": <input
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    type="checkbox"
                    value=""
                  />,
                }
              }
            >
              <RefFindNode
                innerRef={
                  Object {
                    "current": <input
                      class="hidden"
                      readonly=""
                      tabindex="0"
                      type="checkbox"
                      value=""
                    />,
                  }
                }
              >
                <input
                  checked={false}
                  className="hidden"
                  disabled={false}
                  readOnly={true}
                  tabIndex={0}
                  type="checkbox"
                />
              </RefFindNode>
            </Ref>
            <Ref
              innerRef={
                Object {
                  "current": <label />,
                }
              }
            >
              <RefFindNode
                innerRef={
                  Object {
                    "current": <label />,
                  }
                }
              >
                <label />
              </RefFindNode>
            </Ref>
          </div>
        </Checkbox>
      </div>
    </Input>
  </InputController>
</Checkbox>
`;

exports[`<Checkbox /> if \`props.isLabelLeft\` is true should return the right structure 1`] = `
<Checkbox
  error={false}
  isCompact={false}
  isDisabled={false}
  isFluid={false}
  isLabelLeft={true}
  isRadioButton={false}
  isToggle={false}
  label=""
  name=""
  onChange={[Function]}
  onClick={[Function]}
>
  <InputController
    adaptOnChangeEvent={[Function]}
    error={false}
    icon={null}
    initialValue=""
    inputOnChangeFunctionName="onChange"
    isCompact={false}
    isFluid={false}
    isFocused={false}
    isValid={false}
    label={null}
    mapValueToProps={[Function]}
    name=""
    onChange={[Function]}
  >
    <Input
      className=""
      fluid={false}
      iconPosition={null}
      type="text"
    >
      <div
        className="ui input"
      >
        <Checkbox
          checked={false}
          className="is-label-left"
          disabled={false}
          key=".0"
          label=""
          onChange={[Function]}
          onClick={[Function]}
          radio={false}
          toggle={false}
          type="checkbox"
        >
          <div
            className="ui checkbox is-label-left"
            onChange={[Function]}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
          >
            <Ref
              innerRef={
                Object {
                  "current": <input
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    type="checkbox"
                    value=""
                  />,
                }
              }
            >
              <RefFindNode
                innerRef={
                  Object {
                    "current": <input
                      class="hidden"
                      readonly=""
                      tabindex="0"
                      type="checkbox"
                      value=""
                    />,
                  }
                }
              >
                <input
                  checked={false}
                  className="hidden"
                  disabled={false}
                  readOnly={true}
                  tabIndex={0}
                  type="checkbox"
                />
              </RefFindNode>
            </Ref>
            <Ref
              innerRef={
                Object {
                  "current": <label />,
                }
              }
            >
              <RefFindNode
                innerRef={
                  Object {
                    "current": <label />,
                  }
                }
              >
                <label />
              </RefFindNode>
            </Ref>
          </div>
        </Checkbox>
      </div>
    </Input>
  </InputController>
</Checkbox>
`;

exports[`<Checkbox /> should return the right structure 1`] = `
<Checkbox
  error={false}
  isCompact={false}
  isDisabled={false}
  isFluid={false}
  isLabelLeft={false}
  isRadioButton={false}
  isToggle={false}
  label=""
  name=""
  onChange={[Function]}
  onClick={[Function]}
>
  <InputController
    adaptOnChangeEvent={[Function]}
    error={false}
    icon={null}
    initialValue=""
    inputOnChangeFunctionName="onChange"
    isCompact={false}
    isFluid={false}
    isFocused={false}
    isValid={false}
    label={null}
    mapValueToProps={[Function]}
    name=""
    onChange={[Function]}
  >
    <Input
      className=""
      fluid={false}
      iconPosition={null}
      type="text"
    >
      <div
        className="ui input"
      >
        <Checkbox
          checked={false}
          className=""
          disabled={false}
          key=".0"
          label=""
          onChange={[Function]}
          onClick={[Function]}
          radio={false}
          toggle={false}
          type="checkbox"
        >
          <div
            className="ui checkbox"
            onChange={[Function]}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
          >
            <Ref
              innerRef={
                Object {
                  "current": <input
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    type="checkbox"
                    value=""
                  />,
                }
              }
            >
              <RefFindNode
                innerRef={
                  Object {
                    "current": <input
                      class="hidden"
                      readonly=""
                      tabindex="0"
                      type="checkbox"
                      value=""
                    />,
                  }
                }
              >
                <input
                  checked={false}
                  className="hidden"
                  disabled={false}
                  readOnly={true}
                  tabIndex={0}
                  type="checkbox"
                />
              </RefFindNode>
            </Ref>
            <Ref
              innerRef={
                Object {
                  "current": <label />,
                }
              }
            >
              <RefFindNode
                innerRef={
                  Object {
                    "current": <label />,
                  }
                }
              >
                <label />
              </RefFindNode>
            </Ref>
          </div>
        </Checkbox>
      </div>
    </Input>
  </InputController>
</Checkbox>
`;
